### PR TITLE
Record parent build number reliably

### PR DESCRIPTION
**Pull request by olivergondza**
```
Feb 28, 2017 8:59:56 AM hudson.matrix.MatrixConfiguration newBuild
WARNING: guessing that the correct build of PROJECT_ID is #19
Feb 28, 2017 8:59:56 AM hudson.model.Executor run
SEVERE: Unexpected executor death
java.lang.IllegalStateException: PROJECT_PATH/builds/19 already existed; will not overwite with PROJECT_ID #19
	at hudson.model.RunMap.put(RunMap.java:187)
	at hudson.matrix.MatrixConfiguration.newBuild(MatrixConfiguration.java:296)
	at hudson.matrix.MatrixConfiguration.newBuild(MatrixConfiguration.java:75)
	at hudson.model.AbstractProject.createExecutable(AbstractProject.java:1209)
	at hudson.model.AbstractProject.createExecutable(AbstractProject.java:144)
	at hudson.model.Executor$1.call(Executor.java:362)
	at hudson.model.Executor$1.call(Executor.java:344)
	at hudson.model.Queue._withLock(Queue.java:1348)
	at hudson.model.Queue.withLock(Queue.java:1213)
	at hudson.model.Executor.run(Executor.java:344)
```

### CODE REVIEW

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 1]
```diff
@@ -497,18 +496,27 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
+            }
+            number = parent == null ? null : parent.getNumber();
         }
-        
+
         public boolean shouldSchedule(List<Action> actions) {
             return true;
         }
+
+        public MatrixBuild resolve(MatrixProject p) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: Confusing public API method.
Maybe `@CheckForNull MatrixBuild resolveParentBuild(...)` ? + some Javadoc

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 2]
```diff
@@ -497,18 +496,27 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
+
+        private final Integer number;
         public transient MatrixBuild parent;
 
         public ParentBuildAction() {
             final Executor currentExecutor = Executor.currentExecutor();
-            this.parent = currentExecutor != null 
-                    ? (MatrixBuild)currentExecutor.getCurrentExecutable() : null;
+            if (currentExecutor != null) {
+                parent = (MatrixBuild) currentExecutor.getCurrentExecutable();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: Unchecked cast in the old code?

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 3]
```diff
@@ -497,18 +496,27 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
         }
-        
+
         public boolean shouldSchedule(List<Action> actions) {
             return true;
         }
+
+        public MatrixBuild resolve(MatrixProject p) {
+            if (parent != null) return parent;
+            if (number != null) return p.getBuildByNumber(number);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: I suppose the result should be cached in `parent`

**Comment by olivergondza** [comment]
@oleg-nenashev, adjusted. thanks

**Comment by olivergondza** [comment]
I have just noticed this is already implemented in #31 in a better way. Closing this.

**Pull request closed without merging**